### PR TITLE
fix: do not filter out rows where PARTITION BY resolves to null

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 

Note: stacked on top of https://github.com/confluentinc/ksql/pull/4816, (which is now merged)

Fixes: #4747

This commit removes the filter that was excluding any rows where the `PARTITION BY` clause resolved to a `null` value, i.e. either because the result was `null` or because an error occurred evaluating the expression.

This change will only affect new queries started. Pre-existing queries will continue to run as before.

### Testing done 

QTT tests added.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

**Comment by big-andy-coates** [comment]
FYI, the docs call out the limitations of partitioning by a null, and suggest using `COALESCE` from https://github.com/confluentinc/ksql/pull/4829 to avoid.

File: docs-md/developer-guide/joins/partition-data.md [hunk 1]
```diff
@@ -153,6 +153,12 @@ only if they are also both in the same partition after the repartition.
 Otherwise, Kafka is likely to interleave messages. The use case will determine
 if these ordering guarantees are acceptable.
 
+!!! important
+      If the PARTITION BY expression evaluates to NULL the resulting row will be produced to a
```
Comment (JimGalasyn) [inline docs-md/developer-guide/joins/partition-data.md]:
```suggestion
      If the PARTITION BY expression evaluates to NULL, the resulting row is produced to a
```

File: docs-md/developer-guide/joins/partition-data.md [hunk 2]
```diff
@@ -153,6 +153,12 @@ only if they are also both in the same partition after the repartition.
...
+      random partition. You many want to use [COALESCE](../syntax-reference#coalesce) to wrap
+      the expression convert any NULL values to a default value, e.g.
```
Comment (JimGalasyn) [inline docs-md/developer-guide/joins/partition-data.md]:
```suggestion
      the expression and convert any NULL values to a default value, for example,
```

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/joins.json [hunk 3]
```diff
@@ -257,14 +257,14 @@
...
-        {"topic": "_confluent-ksql-some.ksql.service.idquery_CSAS_LEFT_OUTER_JOIN_0-KSTREAM-JOINTHIS-0000000018-store-changelog", "window": {"start": 17000, "end": 28000, "type": "time"}, "key": 90, "value": {"T_ROWTIME": 17000, "T_K": "", "T_ID": 90, "T_NAME": "ninety", "T_VALUE": 90}, "timestamp": 17000},
-        {"topic": "_confluent-ksql-some.ksql.service.idquery_CSAS_LEFT_OUTER_JOIN_0-KSTREAM-JOINTHIS-0000000018-store-changelog", "window": {"start": 30000, "end": 41000, "type": "time"}, "key": 0, "value": {"T_ROWTIME": 30000, "T_K": "", "T_ID": 0, "T_NAME": "bar", "T_VALUE": 99}, "timestamp": 30000},
+        {"topic": "_confluent-ksql-some.ksql.service.idquery_CSAS_LEFT_OUTER_JOIN_0-KSTREAM-JOINTHIS-0000000016-store-changelog", "window": {"start": 0, "end": 11000, "type": "time"}, "key": 0, "value": {"T_ROWTIME": 0, "T_K": "", "T_ID": 0, "T_NAME": "zero", "T_VALUE": 0}, "timestamp": 0},
```
Comment (big-andy-coates) [inline ksqldb-functional-tests/src/test/resources/query-validation-tests/joins.json]: Note: it's safe to change these as they are from an unreleased feature...

**Pull request merged**